File: frontend/src/api/sms.ts
/**
 * SMS API Client
 *
 * This module provides a client for managing SMS communications with candidates,
 * including sending SMS messages, tracking delivery status, and retrieving SMS history.
 *
 * @example
 * ```ts
 * import { smsClient } from '@/api/sms';
 *
 * // Send an SMS message
 * const sms = await smsClient.send({
 *   candidate_id: 'resume-123',
 *   to_number: '+1234567890',
 *   message: 'Hello! We would like to schedule an interview.',
 *   provider: 'Twilio',
 *   recruiter_id: 'recruiter-123'
 * });
 *
 * // List SMS messages for a candidate
 * const messages = await smsClient.list('resume-123');
 *
 * // Get delivery status
 * const status = await smsClient.getDeliveryStatus(undefined, 'sms-id');
 * ```
 */

import axios, { AxiosInstance, AxiosError, AxiosResponse } from 'axios';
import type { ApiError } from '@/types/api';

/**
 * SMS send request
 */
export interface SMSSendRequest {
  candidate_id: string;
  recruiter_id?: string;
  to_number: string;
  from_number?: string;
  message: string;
  provider: string;
  vacancy_id?: string;
}

/**
 * SMS response
 */
export interface SMSResponse {
  id: string;
  communication_id: string;
  candidate_id: string;
  recruiter_id: string | null;
  to_number: string;
  from_number: string | null;
  message: string;
  provider: string;
  delivery_status: string;
  delivery_error: string | null;
  provider_message_id: string | null;
  segment_count: number | null;
  sent_at: string | null;
  created_at: string;
  updated_at: string;
}

/**
 * SMS list response
 */
export interface SMSListResponse {
  messages: SMSResponse[];
  total_count: number;
}

/**
 * Delivery status response
 */
export interface DeliveryStatusResponse {
  id: string;
  delivery_status: string;
  provider_message_id: string | null;
  delivery_error: string | null;
  sent_at: string | null;
  updated_at: string;
}

/**
 * Default API configuration for SMS client
 */
const DEFAULT_CONFIG = {
  baseURL: import.meta.env.VITE_API_URL ?? '',
  timeout: 10000, // 10 seconds
  headers: {
    'Content-Type': 'application/json',
  },
};

/**
 * SMS API Client class
 *
 * Provides methods for managing SMS communications with proper
 * error handling and type safety.
 */
export class SMSClient {
  private client: AxiosInstance;

  /**
   * Create a new SMS client instance
   *
   * @param config - Optional configuration overrides
   */
  constructor(config: Partial<typeof DEFAULT_CONFIG> = {}) {
    const finalConfig = { ...DEFAULT_CONFIG, ...config };

    this.client = axios.create(finalConfig);

    // Response interceptor for error handling
    this.client.interceptors.response.use(
      (response) => response,
      (error) => Promise.reject(this.transformError(error))
    );
  }

  /**
   * Transform Axios error to standardized API error
   *
   * @param error - Axios error
   * @returns Transformed API error
   */
  private transformError(error: unknown): ApiError {
    const axiosError = error as AxiosError<{ detail?: string }>;

    // Network error (no response)
    if (!axiosError.response) {
      if (axiosError.code === 'ECONNABORTED') {
        return {
          detail: 'Request timeout. Please check your connection and try again.',
          status: 408,
        };
      }
      return {
        detail: 'Network error. Please check your connection and try again.',
        status: 0,
      };
    }

    // Server returned error response
    const status = axiosError.response.status;
    const data = axiosError.response.data;

    // Use server's error message if available
    if (data?.detail) {
      return { detail: data.detail, status };
    }

    // Default error messages by status code
    const defaultMessages: Record<number, string> = {
      400: 'Invalid request. Please check your input.',
      401: 'Unauthorized. Please log in.',
      403: 'Forbidden. You do not have permission.',
      404: 'Resource not found.',
      409: 'A conflict occurred with the existing SMS.',
      422: 'Validation error. Please check your input.',
      429: 'Too many requests. Please try again later.',
      500: 'Server error. Please try again later.',
      502: 'Bad gateway. Please try again later.',
      503: 'Service unavailable. Please try again later.',
    };

    return {
      detail: data?.detail || defaultMessages[status] || 'An unexpected error occurred.',
      status,
    };
  }

  /**
   * Send an SMS message to a candidate
   *
   * @param request - Send request with SMS details
   * @returns Sent SMS message details
   * @throws ApiError if sending fails
   *
   * @example
   * ```ts
   * const sms = await smsClient.send({
   *   candidate_id: 'resume-123',
   *   to_number: '+1234567890',
   *   message: 'Hello! We would like to schedule an interview.',
   *   provider: 'Twilio',
   *   recruiter_id: 'recruiter-123'
   * });
   * ```
   */
  async send(request: SMSSendRequest): Promise<SMSResponse> {
    try {
      const response: AxiosResponse<SMSResponse> = await this.client.post(
        '/api/sms/send',
        request
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * List SMS messages with optional filters
   *
   * @param candidateId - Optional candidate ID filter
   * @param recruiterId - Optional recruiter ID filter
   * @param provider - Optional SMS provider filter
   * @param deliveryStatus - Optional delivery status filter
   * @param limit - Maximum number of results to return (default: 100)
   * @param offset - Number of results to skip (default: 0)
   * @returns List of SMS messages
   * @throws ApiError if listing fails
   *
   * @example
   * ```ts
   * // Get all SMS messages for a candidate
   * const messages = await smsClient.list('resume-123');
   *
   * // Get only delivered SMS messages
   * const delivered = await smsClient.list(
   *   'resume-123',
   *   undefined,
   *   undefined,
   *   'delivered'
   * );
   *
   * // Get SMS messages from a specific provider
   * const twilioSms = await smsClient.list(
   *   undefined,
   *   undefined,
   *   'Twilio'
   * );
   * ```
   */
  async list(
    candidateId?: string,
    recruiterId?: string,
    provider?: string,
    deliveryStatus?: string,
    limit = 100,
    offset = 0
  ): Promise<SMSListResponse> {
    try {
      const params: Record<string, string | number> = {};
      if (candidateId) params.candidate_id = candidateId;
      if (recruiterId) params.recruiter_id = recruiterId;
      if (provider) params.provider = provider;
      if (deliveryStatus) params.delivery_status = deliveryStatus;
      params.limit = limit;
      params.offset = offset;

      const response: AxiosResponse<SMSListResponse> = await this.client.get(
        '/api/sms/',
        { params }
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Get a specific SMS message by ID
   *
   * @param id - SMS message ID
   * @returns SMS message details
   * @throws ApiError if retrieval fails
   *
   * @example
   * ```ts
   * const sms = await smsClient.get('sms-id');
   * ```
   */
  async get(id: string): Promise<SMSResponse> {
    try {
      const response: AxiosResponse<SMSResponse> = await this.client.get(
        `/api/sms/${id}`
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Get delivery status for an SMS message
   *
   * @param providerMessageId - Optional provider message ID to query
   * @param smsId - Optional SMS ID to query
   * @returns Delivery status information
   * @throws ApiError if retrieval fails
   *
   * @example
   * ```ts
   * // Query by SMS ID
   * const status = await smsClient.getDeliveryStatus(undefined, 'sms-id');
   *
   * // Query by provider message ID
   * const status = await smsClient.getDeliveryStatus('provider-msg-id');
   * ```
   */
  async getDeliveryStatus(
    providerMessageId?: string,
    smsId?: string
  ): Promise<DeliveryStatusResponse> {
    try {
      const params: Record<string, string> = {};
      if (providerMessageId) params.provider_message_id = providerMessageId;
      if (smsId) params.sms_id = smsId;

      const response: AxiosResponse<DeliveryStatusResponse> = await this.client.get(
        '/api/sms/delivery-status',
        { params }
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }
}

/**
 * Default SMS client instance
 */
export const smsClient = new SMSClient();

File: backend/api/sms.py
"""
SMS message management endpoints.

This module provides endpoints for managing SMS communications with candidates,
including sending SMS messages, tracking delivery status, and retrieving SMS history.
Supports multiple SMS providers (Twilio, AWS SNS, etc.) with delivery tracking.
"""
import logging
from typing import List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from database import get_db
from models.communication import Communication, CommunicationDirection, CommunicationStatus, CommunicationType
from models.sms_message import SMSMessage, SMSDeliveryStatus

logger = logging.getLogger(__name__)

router = APIRouter()


class SMSSendRequest(BaseModel):
    """Request model for sending an SMS message."""

    candidate_id: str = Field(..., description="Candidate ID (resume ID) to send SMS to")
    recruiter_id: Optional[str] = Field(None, description="Recruiter ID (sender) of the SMS")
    to_number: str = Field(..., min_length=10, max_length=50, description="Recipient phone number (E.164 format)")
    from_number: Optional[str] = Field(None, min_length=10, max_length=50, description="Sender phone number (E.164 format)")
    message: str = Field(..., min_length=1, max_length=1600, description="SMS message content")
    provider: str = Field(..., min_length=1, max_length=100, description="SMS provider (e.g., Twilio, AWS SNS)")
    vacancy_id: Optional[str] = Field(None, description="Optional vacancy ID associated with this SMS")


class SMSResponse(BaseModel):
    """Response model for a single SMS message."""

    id: str = Field(..., description="Unique identifier for the SMS")
    communication_id: str = Field(..., description="Communication ID")
    candidate_id: str = Field(..., description="Candidate ID (recipient)")
    recruiter_id: Optional[str] = Field(None, description="Recruiter ID (sender)")
    to_number: str = Field(..., description="Recipient phone number")
    from_number: Optional[str] = Field(None, description="Sender phone number")
    message: str = Field(..., description="SMS message content")
    provider: str = Field(..., description="SMS provider")
    delivery_status: str = Field(..., description="Delivery status")
    delivery_error: Optional[str] = Field(None, description="Error message if delivery failed")
    provider_message_id: Optional[str] = Field(None, description="Provider's message ID")
    segment_count: Optional[int] = Field(None, description="Number of SMS segments used")
    sent_at: Optional[str] = Field(None, description="Timestamp when SMS was sent")
    created_at: str = Field(..., description="Creation timestamp")
    updated_at: str = Field(..., description="Last update timestamp")


class SMSListResponse(BaseModel):
    """Response model for listing SMS messages."""

    messages: List[SMSResponse] = Field(..., description="List of SMS messages")
    total_count: int = Field(..., description="Total number of SMS messages")


class DeliveryStatusResponse(BaseModel):
    """Response model for delivery status query."""

    id: str = Field(..., description="SMS ID")
    delivery_status: str = Field(..., description="Current delivery status")
    provider_message_id: Optional[str] = Field(None, description="Provider's message ID")
    delivery_error: Optional[str] = Field(None, description="Error message if delivery failed")
    sent_at: Optional[str] = Field(None, description="Timestamp when SMS was sent")
    updated_at: str = Field(..., description="Last update timestamp")


@router.post(
    "/send",
    response_model=SMSResponse,
    status_code=status.HTTP_201_CREATED,
    tags=["SMS"],
)
async def send_sms(
    request: SMSSendRequest,
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """
    Send an SMS message to a candidate.

    This endpoint creates and sends an SMS message to a candidate. The SMS is
    logged in the database with delivery tracking. The actual sending is handled
    by a background worker (Celery task) that integrates with the specified SMS provider.

    Args:
        request: Request body containing SMS details
        db: Database session

    Returns:
        JSON response with sent SMS details

    Raises:
        HTTPException(404): If candidate is not found
        HTTPException(422): If validation fails
        HTTPException(500): If an internal error occurs

    Examples:
        >>> import requests
        >>> response = requests.post(
        ...     "http://localhost:8000/api/sms/send",
        ...     json={
        ...         "candidate_id": "resume-uuid",
        ...         "to_number": "+1234567890",
        ...         "message": "Hello! We'd like to schedule an interview.",
        ...         "provider": "Twilio",
        ...         "recruiter_id": "recruiter-uuid"
        ...     }
        ... )
        >>> response.status_code
        201
    """
    try:
        logger.info(f"Sending SMS to candidate: {request.candidate_id}")

        # Verify candidate exists
        from models.resume import Resume

        candidate_result = await db.execute(
            select(Resume).where(Resume.id == UUID(request.candidate_id))
        )
        candidate = candidate_result.scalar_one_or_none()

        if not candidate:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Candidate not found: {request.candidate_id}",
            )

        # Calculate segment count (standard SMS is 160 chars, multipart is 153 chars per segment)
        segment_count = (len(request.message) + 152) // 153 if len(request.message) > 160 else 1

        # Create communication record
        communication = Communication(
            candidate_id=UUID(request.candidate_id),
            recruiter_id=UUID(request.recruiter_id) if request.recruiter_id else None,
            vacancy_id=UUID(request.vacancy_id) if request.vacancy_id else None,
            type=CommunicationType.SMS,
            direction=CommunicationDirection.OUTBOUND,
            status=CommunicationStatus.PENDING,
            subject=None,  # SMS doesn't have subject
            body=request.message,
        )
        db.add(communication)
        await db.flush()

        # Create SMS message record
        new_sms = SMSMessage(
            communication_id=communication.id,
            to_number=request.to_number,
            from_number=request.from_number,
            provider=request.provider,
            delivery_status=SMSDeliveryStatus.PENDING,
            segment_count=segment_count,
        )
        db.add(new_sms)
        await db.flush()

        # Prepare response data
        response_data = {
            "id": str(new_sms.id),
            "communication_id": str(new_sms.communication_id),
            "candidate_id": str(communication.candidate_id),
            "recruiter_id": str(communication.recruiter_id) if communication.recruiter_id else None,
            "to_number": new_sms.to_number,
            "from_number": new_sms.from_number,
            "message": communication.body,
            "provider": new_sms.provider,
            "delivery_status": new_sms.delivery_status.value,
            "delivery_error": new_sms.delivery_error,
            "provider_message_id": new_sms.provider_message_id,
            "segment_count": new_sms.segment_count,
            "sent_at": communication.sent_at.isoformat() if communication.sent_at else None,
            "created_at": new_sms.created_at.isoformat(),
            "updated_at": new_sms.updated_at.isoformat(),
        }

        await db.commit()

        logger.info(f"SMS created with ID: {new_sms.id}, queued for sending")

        # TODO: Trigger Celery task to send SMS via provider
        # from tasks.sms_task import send_sms_task
        # send_sms_task.delay(str(new_sms.id))

        return JSONResponse(
            status_code=status.HTTP_201_CREATED,
            content=response_data,
        )

    except HTTPException:
        raise
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Invalid UUID format",
        )
    except Exception as e:
        logger.error(f"Error sending SMS: {e}", exc_info=True)
        await db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to send SMS: {str(e)}",
        ) from e


@router.get("/", tags=["SMS"])
async def list_sms_messages(
    candidate_id: Optional[str] = Query(None, description="Filter by candidate ID"),
    recruiter_id: Optional[str] = Query(None, description="Filter by recruiter ID"),
    provider: Optional[str] = Query(None, description="Filter by SMS provider"),
    delivery_status: Optional[str] = Query(None, description="Filter by delivery status"),
    limit: int = Query(100, ge=1, le=1000, description="Maximum number of results"),
    offset: int = Query(0, ge=0, description="Number of results to skip"),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """
    List SMS messages with optional filters.

    This endpoint retrieves SMS messages with support for filtering
    by candidate, recruiter, provider, and delivery status.

    Args:
        candidate_id: Optional candidate ID filter
        recruiter_id: Optional recruiter ID filter
        provider: Optional provider filter
        delivery_status: Optional delivery status filter
        limit: Maximum number of results to return (default: 100)
        offset: Number of results to skip (default: 0)
        db: Database session

    Returns:
        JSON response with list of SMS messages

    Raises:
        HTTPException(500): If an internal error occurs

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/sms/?candidate_id=resume-uuid")
        >>> response.json()
        {
            "messages": [...],
            "total_count": 5
        }
    """
    try:
        logger.info(f"Listing SMS messages with filters: candidate_id={candidate_id}, provider={provider}")

        # Build base query
        query = (
            select(SMSMessage, Communication)
            .join(Communication, SMSMessage.communication_id == Communication.id)
            .where(Communication.type == CommunicationType.SMS)
        )

        # Apply filters
        if candidate_id:
            query = query.where(Communication.candidate_id == UUID(candidate_id))
        if recruiter_id:
            query = query.where(Communication.recruiter_id == UUID(recruiter_id))
        if provider:
            query = query.where(SMSMessage.provider == provider)
        if delivery_status:
            try:
                status_enum = SMSDeliveryStatus(delivery_status)
                query = query.where(SMSMessage.delivery_status == status_enum)
            except ValueError:
                raise HTTPException(
                    status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                    detail=f"Invalid delivery status: {delivery_status}",
                )

        # Get total count
        count_query = select(SMSMessage).join(
            Communication, SMSMessage.communication_id == Communication.id
        )
        if candidate_id:
            count_query = count_query.where(Communication.candidate_id == UUID(candidate_id))
        if recruiter_id:
            count_query = count_query.where(Communication.recruiter_id == UUID(recruiter_id))
        if provider:
            count_query = count_query.where(SMSMessage.provider == provider)
        if delivery_status:
            count_query = count_query.where(SMSMessage.delivery_status == status_enum)

        count_result = await db.execute(select(func.count()).select_from(count_query.subquery()))
        total_count = count_result.scalar() or 0

        # Apply pagination and ordering
        query = query.order_by(SMSMessage.created_at.desc()).offset(offset).limit(limit)

        result = await db.execute(query)
        rows = result.all()

        messages = []
        for sms, communication in rows:
            messages.append({
                "id": str(sms.id),
                "communication_id": str(sms.communication_id),
                "candidate_id": str(communication.candidate_id),
                "recruiter_id": str(communication.recruiter_id) if communication.recruiter_id else None,
                "to_number": sms.to_number,
                "from_number": sms.from_number,
                "message": communication.body or "",
                "provider": sms.provider,
                "delivery_status": sms.delivery_status.value,
                "delivery_error": sms.delivery_error,
                "provider_message_id": sms.provider_message_id,
                "segment_count": sms.segment_count,
                "sent_at": communication.sent_at.isoformat() if communication.sent_at else None,
                "created_at": sms.created_at.isoformat(),
                "updated_at": sms.updated_at.isoformat(),
            })

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "messages": messages,
                "total_count": total_count,
            },
        )

    except HTTPException:
        raise
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Invalid UUID format",
        )
    except Exception as e:
        logger.error(f"Error listing SMS messages: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to list SMS messages: {str(e)}",
        ) from e


@router.get("/{sms_id}", tags=["SMS"])
async def get_sms(
    sms_id: str,
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """
    Get a single SMS message by ID.

    This endpoint retrieves details of a specific SMS message including
    delivery status and provider information.

    Args:
        sms_id: SMS ID (UUID)
        db: Database session

    Returns:
        JSON response with SMS details

    Raises:
        HTTPException(404): If SMS is not found
        HTTPException(422): If UUID format is invalid
        HTTPException(500): If an internal error occurs

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/sms/sms-uuid")
        >>> response.json()
        {
            "id": "sms-uuid",
            "communication_id": "...",
            "to_number": "+1234567890",
            "message": "Hello!",
            "delivery_status": "delivered"
        }
    """
    try:
        logger.info(f"Fetching SMS message: {sms_id}")

        result = await db.execute(
            select(SMSMessage, Communication)
            .join(Communication, SMSMessage.communication_id == Communication.id)
            .where(SMSMessage.id == UUID(sms_id))
        )
        row = result.first()

        if not row:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"SMS message not found: {sms_id}",
            )

        sms, communication = row

        response_data = {
            "id": str(sms.id),
            "communication_id": str(sms.communication_id),
            "candidate_id": str(communication.candidate_id),
            "recruiter_id": str(communication.recruiter_id) if communication.recruiter_id else None,
            "to_number": sms.to_number,
            "from_number": sms.from_number,
            "message": communication.body or "",
            "provider": sms.provider,
            "delivery_status": sms.delivery_status.value,
            "delivery_error": sms.delivery_error,
            "provider_message_id": sms.provider_message_id,
            "segment_count": sms.segment_count,
            "sent_at": communication.sent_at.isoformat() if communication.sent_at else None,
            "created_at": sms.created_at.isoformat(),
            "updated_at": sms.updated_at.isoformat(),
        }

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except HTTPException:
        raise
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Invalid UUID format",
        )
    except Exception as e:
        logger.error(f"Error fetching SMS message: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch SMS message: {str(e)}",
        ) from e


@router.get("/delivery-status", tags=["SMS"])
async def get_delivery_status(
    provider_message_id: Optional[str] = Query(None, description="Provider message ID to query"),
    sms_id: Optional[str] = Query(None, description="SMS ID to query"),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """
    Get delivery status for SMS messages.

    This endpoint retrieves the current delivery status of SMS messages.
    You can query by either the provider's message ID or the SMS ID.

    Args:
        provider_message_id: Optional provider message ID filter
        sms_id: Optional SMS ID filter
        db: Database session

    Returns:
        JSON response with delivery status information

    Raises:
        HTTPException(422): If neither filter is provided or both are provided
        HTTPException(404): If SMS is not found
        HTTPException(500): If an internal error occurs

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/sms/delivery-status?sms_id=sms-uuid")
        >>> response.json()
        {
            "id": "sms-uuid",
            "delivery_status": "delivered",
            "provider_message_id": "provider-msg-id"
        }
    """
    try:
        # Validate that exactly one filter is provided
        if provider_message_id and sms_id:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail="Provide either provider_message_id or sms_id, not both",
            )
        if not provider_message_id and not sms_id:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail="Provide either provider_message_id or sms_id",
            )

        logger.info(f"Fetching delivery status: sms_id={sms_id}, provider_message_id={provider_message_id}")

        # Build query
        query = select(SMSMessage)
        if sms_id:
            query = query.where(SMSMessage.id == UUID(sms_id))
        else:
            query = query.where(SMSMessage.provider_message_id == provider_message_id)

        result = await db.execute(query)
        sms = result.scalar_one_or_none()

        if not sms:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"SMS message not found",
            )

        # Get communication for sent_at timestamp
        comm_result = await db.execute(
            select(Communication).where(Communication.id == sms.communication_id)
        )
        communication = comm_result.scalar_one_or_none()

        response_data = {
            "id": str(sms.id),
            "delivery_status": sms.delivery_status.value,
            "provider_message_id": sms.provider_message_id,
            "delivery_error": sms.delivery_error,
            "sent_at": communication.sent_at.isoformat() if communication and communication.sent_at else None,
            "updated_at": sms.updated_at.isoformat(),
        }

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except HTTPException:
        raise
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Invalid UUID format",
        )
    except Exception as e:
        logger.error(f"Error fetching delivery status: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch delivery status: {str(e)}",
        ) from e
